enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,378 @@
+- In particular, using CVO and having it be a day 1 install operation meets the ease of use targets we are trying to
+reach with this proposal
+- Crafting OLM dependencies between this proposal and [related proposals](https://github.com/openshift/enhancements/pull/214)
+as part of an OLM based install at the time of this writing is immature.
+
+#### An overview of CSI and the main K8S hook points
+
+ - a [CSI plugin](https://github.com/container-storage-interface/spec) is a gRPC endpoint that implements CSI services and
+ allows storage vendors to develop a storage solution that works across a number of container orchestration systems.
+ - though for our purposes, we will not provide a generic CSI plugin, but rather a special case one that is only intended
+ to work on OpenShift
+ - also for our purposes, `plugin`, `driver`, and `endpoint` are the same thing based on which document you are
+ reading.
+ - k8s provides Kubernetes CSI Sidecar Containers: a set of standard containers that aim to simplify the development and 
+ deployment of CSI Drivers on Kubernetes.
+ - There are CSI volume and persistent volume types in k8s.
+ - per [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#csi) "the CSI types do not support direct
+ reference from a Pod and may only be referenced in a Pod via a `PersistentVolumeClaim` object" (direct quote)
+ - However, @gnufied tells us that documentation is out of date and incorrect.  To quote him:  That is where [inline 
+ CSI volumes come](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html).  @jsafrane also echoed this.
+ - To that end, one of the sidecar containers, the [external provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html),
+ helps with this
+ - It watches Kubernetes `PersistentVolumeClaim` objects, and if the PVC 
+ references a Kubernetes StorageClass, and the name in the provisioner field of the storage class matches the name 
+ returned by the specified CSI endpoint `GetPluginInfo` call, it calls the CSI endpoint `CreateVolume` method to provision
+ the new volume.
+ - there are also lifecycle methods for destroy, resize, etc.
+ - On the more recently introduced inline ephemeral mode of which @jsafrane has strongly suggested to us and which 
+ @gnufied also noted ... 
+ - In the ephemeral case, the Pod spec lists a volume and is considered to originate or "inline" it.
+ - The `csi` subfield is used, and the storage provider's name is listed as with the PVC based approach.
+ 
+ ```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: some-pod
+spec:
+  containers:
+    ...
+     volumeMounts:
+        - name: entitlement
+          mountPath: /etc/entitlement
+  volumes:
+      - name: entitlement
+        csi:
+          driver: secretprovider.openshift.io
+          volumeAttributes:
+              namespace: openshift-config
```

Comment:
I would not expect an API shaped like this.

Reply:
After going over this comment a few times (bouncing between a few interpretations), and cross referencing with your `ProjectedResource` suggestion below, I'm landing with this
- in order to hide the originating coordinates as you suggest below, a reference to the namespace the shared secret is in, and the shared secret name itself, should **NOT** be present in the k8s CSI volume definition in this example
- instead, a reference to `ProjectedResource` should exist in the Pod (presumably an annotation, but you didn't previously specify, so please advise further if needed)
- the admission plugin will decide based on the ACL checks if the pod's SA is even allowed to use said `ProjectResource`
- if so, it will mutate the pod to add the k8s CSI volume def, similar to this example, but with a notable exception
- namely, in that k8s CSI volume def, the volumeAttributes would list the requested `ProjectedResource`
- the CSI driver will look up the `ProjectedResrouce`, then look up the referenced Secret/Configmap, and go from there

Sound close @deads2k ?

Reply:
```
csi:
  driver: projectedresource.storage.openshift.io
    volumeAttributes:
      projectedResourceName: the-projected-resource
```

this way the information provided matches the projectedresource (clusterscoped) and cannot be used for user-defined fishing of namespaces.